### PR TITLE
support checking paste in signaler functions

### PR DESCRIPTION
**Pull request by LiNk-NY**
If you are submitting a pull request to `BiocCheck` please follow the instructions outlined in [this presentation](https://docs.google.com/presentation/d/1DkN2WVPOMVGqUtlSSrWbx6IMjtGP_cEHoE3nfOEnD68/edit#slide=id.p). This presentation includes steps for forking, creating a branch for you to work on, and useful related information.

Prior to sending the pull request, please verify that `R CMD build` and `R CMD check` run without warnings or errors on the latest Bioconductor-devel and complete the following steps:

* [x] Update the NEWS file (required)
* [x] Update the vignette file (required)
* [x] Add unit tests (optional)

This flags any `paste0` and `paste` function calls within an enclosing `message`, `warning` or `stop` function call as a
*NOTE*.

### CODE REVIEW

File: R/checks.R [hunk 1]
```diff
@@ -1071,6 +1080,30 @@ checkSingleColon <- function(Rdir, avail_pkgs = character(0L)) {
     msg_sc <- unlist(msg_sc)
 }
 
+checkPasteInSignaler <- function(Rdir) {
+    rfiles <- dir(Rdir, pattern = "\\.[Rr]$", full.names = TRUE)
+    names(rfiles) <- basename(rfiles)
+    paste_sig <- lapply(rfiles, function(rfile) {
```
Comment (mtmorgan) [inline R/checks.R]: Maybe make this anonymous function a named function, to help with readability and testing? Otherwise this looks like a great addition!

**Review by mtmorgan — APPROVED** [review]
Looks great, with the tweak at line 1086 as sugggested

File: R/checks.R (hunk 1 above)
Comment (LiNk-NY) [reply R/checks.R]: https://github.com/Bioconductor/BiocCheck/pull/111/files#diff-6cd987920443d42098a4963d43b2e0efec9fd16f9198d66705e6d95094e3b87aR1083-R1100

**Comment by LiNk-NY** [comment]
Thanks!

**Comment by hpages** [comment]
Good idea. This is something I see often when I review packages. This and things like:
```
if (...)
    stop("Error: invalid blah blah...")  # will display "Error" twice
```
or
```
if (...) {
    cat("Error: invalid blah blah...")
    return()
}
```
but that's a different one.

I didn't look at the patch but make sure to raise a warning _only_ when `paste()` or `paste0()` are used to concatenate _single_ strings e.g. `paste("Hello", "World")`. Using `paste()` or `paste0()` inside a call to `message()` to combine 2 vectors of length > 1 is not redundant and perfectly legit, and something that is not uncommon.

Also while you are at it, you could add `cat()` and `packageStartupMessage()` to the list.

Thanks,
H.

**Comment by hpages** [comment]
Also you often see things like: `stop("the following genes are not human: ", paste(alien_gene_ids, collapse=", "))`. Perfectly fine too.

**Comment by mtmorgan** [comment]
A unit test would be great for this; does the new helper function make this easy, e.g, 
```
rfile <- tempfile()
writeLines(c(
    "message(paste('foo', 'bar'))",
    "message(paste('foo', x))",
    "message(paste(x, collapse = '\t'))",
    "message('foo', paste(x, collapse = 't'))"
), rfile)
.findPasteInSignaler(rfile)
...
```

**Comment by LiNk-NY** [comment]
> Good idea. This is something I see often when I review packages. This and things like:
> 
> ```
> if (...)
>     stop("Error: invalid blah blah...")  # will display "Error" twice
> ```
> 
> or
> 
> ```
> if (...) {
>     cat("Error: invalid blah blah...")
>     return()
> }
> ```
> 
> but that's a different one.
> 
> I didn't look at the patch but make sure to raise a warning _only_ when `paste()` or `paste0()` are used to concatenate _single_ strings e.g. `paste("Hello", "World")`. Using `paste()` or `paste0()` inside a call to `message()` to combine 2 vectors of length > 1 is not redundant and perfectly legit, and something that is not uncommon.
> 

@hpages Can you open a separate issue for these? Thanks!

> Also while you are at it, you could add `cat()` and `packageStartupMessage()` to the list.

Did you mean these for the new issue(s)?

**Comment by LiNk-NY** [comment]
> A unit test would be great for this; does the new helper function make this easy, e.g,
> 
> ```
> rfile <- tempfile()
> writeLines(c(
>     "message(paste('foo', 'bar'))",
>     "message(paste('foo', x))",
>     "message(paste(x, collapse = '\t'))",
>     "message('foo', paste(x, collapse = 't'))"
> ), rfile)
> .findPasteInSignaler(rfile)
> ...
> ```

@mtmorgan After re-reading this, are we saying that lines 2 and 4 would get a pass?
Currently, 3 and 4 are the exceptions.

**Comment by mtmorgan** [comment]
1 & 2 look unnecessary to me; 3 and 4 look ok.

**Pull request merged**